### PR TITLE
Improved Paged Attention

### PR DESCRIPTION
**Pull request by madamczykhabana**
Rewrite of paged_attention_v1 implementation on gaudi. Blocks are gathered by columns to form a single tensor with whole context, out-of-bounds indices are masked and then regular dot-product attention is calculated.
TODO: add support for alibi and attention_masks

### CODE REVIEW

**Comment by madamczykhabana** [comment]
@kzawora-intel @mdvoretc-intel please review

File: vllm/hpu/ops.py [hunk 1]
```diff
@@ -12,86 +12,49 @@
...
     raise NotImplementedError
 
+
+def fetch_from_cache(cache, blocks):
+    _, seq_len = blocks.shape
+    return torch.cat([cache.index_select(0, blocks[:, i]) for i in range(seq_len)], dim=-1).flatten(0, 1)
+
+
+
 def paged_attention_v1(query_in, key_cache_in, value_cache_in, head_mapping, scale, block_tables, context_lens, block_size, max_context_len, alibi_slopes, attn_masks=None)  -> None:
```
Comment (mdvoretc-intel) [inline vllm/hpu/ops.py:35]: The `_in` suffix is meant to distinguish inputs from the tensors used in the implementation, which are cast to BF16 to ensure the proper functioning of the filler in softmax. Does this version of PA deliver good results on FP16? If no, the type casting may need to be restored. If yes, then the suffixes should be removed.

File: vllm/hpu/ops.py [hunk 2]
```diff
@@ -12,86 +12,49 @@
...
-        out = torch.matmul(attention_probs.to(value.dtype), value).reshape(num_seqs, num_query_heads, head_size)
-        output.add_(out)
     htorch.core.mark_step()
-    return output.to(dtype=query_in.dtype)
+    batch_size, num_head, head_dim = query_in.shape
+    query_in = query_in.view(batch_size * num_head, 1, head_dim)
+    key = fetch_from_cache(key_cache_in, block_tables)
+    value = fetch_from_cache(value_cache_in, block_tables)
+    seq_len = key.size(-1)
+    attn_weights = torch.bmm(query_in, key).mul_(scale)
```
Comment (kzawora-intel) [inline vllm/hpu/ops.py:46]: won't we crash here if num query heads != num kv heads?

Comment (kzawora-intel) [reply vllm/hpu/ops.py:46]: We'd need some additional `repeat_interleave`s for keys and values (i.e. https://github.com/HabanaAI/vllm-fork/pull/4/files#diff-bd4586975c8131c8c80c2103554f75f014ce5d50216126821127a2ea31d96502R255-R259) to handle MQA/GQA cases in LLama70b

File: vllm/hpu/ops.py (hunk 1 above)
Comment (madamczykhabana) [reply vllm/hpu/ops.py:35]: I've double check and it works fine both on fp16 and bf16. I'll remove those suffixes.

File: vllm/hpu/ops.py (hunk 2 above)
Comment (madamczykhabana) [reply vllm/hpu/ops.py:46]: I'll look into it. Thanks!

File: vllm/hpu/ops.py [hunk 3]
```diff
@@ -11,87 +11,72 @@
...
+import vllm.hpu.utils as hpu_utils
+
+
 def silu_and_mul(output, input):
-    htorch.core.mark_step()
     d = input.shape[-1] // 2
     silu = torch.nn.SiLU().to(input.device)
     x, y = torch.split(input, d, dim=-1)
     output.copy_(silu(x) * y)
-    htorch.core.mark_step()
```
Comment (kzawora-intel) [inline vllm/hpu/ops.py:20]: any reason for removing these?

File: vllm/hpu/ops.py (hunk 2 above)
Comment (madamczykhabana) [reply vllm/hpu/ops.py]: I've added support for MQA/GQA. Basic unit tests run fine now. Please let me know if anything else is missing :)

File: vllm/hpu/ops.py (hunk 3 above)
Comment (madamczykhabana) [reply vllm/hpu/ops.py:20]: Basically, not needed. It's a very simple operation. There's no need to break the graph here.

File: vllm/hpu/ops.py (hunk 2 above)
Comment (kzawora-intel) [reply vllm/hpu/ops.py]: Looks good, I was able to get Llama 70B serving working properly with your patch and #16.

**Review by kzawora-intel — APPROVED** [review]

**Review by mdvoretc-intel — APPROVED** [review]
